rust/src/table/convert_to_delta.rs

```diff
@@ -0,0 +1,277 @@
+            let (key, val) = (partition.key.to_string(), partition.value.to_string());
+            debug!(
+                "Found partition {partition:#?} in parquet file {:#?}",
+                file.location
+            );
+            partitions.insert(key.clone(), Some(val));
+            if partition_columns.insert(key.clone()) {
+                partition_schema_fields.push(SchemaField::new(
+                    key,
+                    SchemaDataType::primitive("string".to_string()),
+                    true,
+                    HashMap::new(),
+                ));
+            }
+        }
+    }
+    Ok(partitions)
+}
+
+async fn delta_table(
+    files: Vec<ObjectMeta>,
+    store: Arc<DeltaObjectStore>,
+    url: Url,
+) -> Result<DeltaTable, Error> {
+    let (mut schemas, mut actions, mut partition_columns, mut partition_schema_fields) =
+        (Vec::new(), Vec::new(), HashSet::new(), Vec::new());
+    for file in files {
+        debug!("Processing parquet file: {:#?}", file.location);
+        schemas.push(arrow_schema(&file, store.clone()).await?);
+        let partition_values =
+            partitions(&file, &mut partition_columns, &mut partition_schema_fields)?;
+        actions.push(add_action(&file, partition_values)?);
+    }
+    let mut schema_fields = Schema::try_from(&ArrowSchema::try_merge(schemas)?)?
+        .get_fields()
+        .clone();
+    schema_fields.append(&mut partition_schema_fields);
+    debug!("Schema fields for the parquet table: {schema_fields:#?}");
+    Ok(CreateBuilder::new()
+        .with_location(url)
+        .with_object_store(store.clone())
+        .with_columns(schema_fields)
+        .with_partition_columns(partition_columns.into_iter())
+        .with_actions(actions)
+        .with_save_mode(SaveMode::Ignore)
+        .await?)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        object_store, parquet_files, partitions, url, HashMap, HashSet, SchemaDataType, SchemaField,
+    };
+    use object_store::path::Path;
+    use test_case::test_case;
+
+    fn schema_field(key: &str) -> SchemaField {
+        SchemaField::new(
+            key.to_string(),
+            SchemaDataType::primitive("string".to_string()),
+            true,
+            HashMap::new(),
+        )
+    }
+
+    #[test_case("tests/data/delta-0.8.0" =>
+        vec![
+            Path::from("part-00000-c9b90f86-73e6-46c8-93ba-ff6bfaf892a1-c000.snappy.parquet"),
+            Path::from("part-00000-04ec9591-0b73-459e-8d18-ba5711d6cbe1-c000.snappy.parquet"),
+            Path::from("part-00001-911a94a2-43f6-4acb-8620-5e68c2654989-c000.snappy.parquet"),
+        ]
+    )]
+    #[test_case("tests/data/delta-0.8.0-numeric-partition" =>
+        vec![
+            Path::from("x=9/y=9.9/part-00007-3c50fba1-4264-446c-9c67-d8e24a1ccf83.c000.snappy.parquet"),
+            Path::from("x=10/y=10.0/part-00015-24eb4845-2d25-4448-b3bb-5ed7f12635ab.c000.snappy.parquet"),
+        ]
+    )]
+    #[tokio::test]
```

Comment:
I'm not sure I understand the test_case crate and what this macro is doing that function calls couldn't just do? :confused:

Reply:
Removed the `test_case` macros. Agree that function calls would also work. I used `test_case` because I think it improves the readability of the tests by combining the test function and the function call into one parametrized function.